File: src-tauri/src/domain/repositories/attendance_repository.rs
// Repository Trait: AttendanceRepository
// Abstract interface for attendance data access

use async_trait::async_trait;
use crate::domain::entities::attendance::{Attendance, AttendanceStatus, AttendanceStats};
use crate::domain::errors::{DomainError, DomainResult};

#[async_trait]
pub trait AttendanceRepository: Send + Sync {
    /// Record or update attendance for a student
    async fn record(
        &self,
        student_id: i64,
        class_id: i64,
        date: String,
        status: AttendanceStatus,
        notes: Option<String>,
    ) -> DomainResult<i64>;

    /// Get attendance by ID
    async fn get_by_id(&self, id: i64) -> DomainResult<Attendance>;

    /// Get attendance for a class on a specific date
    async fn get_by_class_and_date(&self, class_id: i64, date: &str) -> DomainResult<Vec<Attendance>>;

    /// Get all unsynced attendance records
    async fn get_unsynced(&self) -> DomainResult<Vec<Attendance>>;

    /// Get unsynced records for a specific class
    async fn get_unsynced_by_class(&self, class_id: i64) -> DomainResult<Vec<Attendance>>;

    /// Mark records as synced
    async fn mark_as_synced(&self, record_ids: Vec<i64>) -> DomainResult<()>;

    /// Get attendance statistics for a class on a specific date
    async fn get_stats(&self, class_id: i64, date: &str) -> DomainResult<AttendanceStats>;

    /// Get attendance for a student within a date range
    async fn get_by_student_and_date_range(
        &self,
        student_id: i64,
        start_date: &str,
        end_date: &str,
    ) -> DomainResult<Vec<Attendance>>;

    /// Check if attendance record exists for a student on a specific date
    async fn exists(&self, student_id: i64, class_id: i64, date: &str) -> DomainResult<bool>;
}

File: flutter/desktop/lib/domain/repositories/attendance_repository.dart
import '../entities/attendance_entity.dart';
import '../core/result.dart';
import '../exceptions/domain_exceptions.dart';

/// Repository interface for attendance-related data operations.
/// 
/// This interface defines the contract for attendance data access without
/// specifying the implementation details, following the Repository pattern.
abstract class AttendanceRepository {
  /// Records attendance for students.
  /// 
  /// [attendanceRecords] is a list of attendance records to save.
  /// Returns the list of saved [AttendanceRecordEntity] on success.
  /// Throws [ValidationException] on invalid data or [DataException] on save failure.
  Future<Result<List<AttendanceRecordEntity>>> recordAttendance(
    List<AttendanceRecordEntity> attendanceRecords,
  );

  /// Retrieves attendance records for a specific class and date.
  /// 
  /// Returns a list of [AttendanceRecordEntity] for the specified class and date.
  Future<Result<List<AttendanceRecordEntity>>> getAttendanceByClassAndDate({
    required int classId,
    required DateTime date,
  });

  /// Retrieves attendance records for a specific student within a date range.
  /// 
  /// Returns a list of [AttendanceRecordEntity] for the specified student.
  Future<Result<List<AttendanceRecordEntity>>> getAttendanceByStudentAndDateRange({
    required int studentId,
    required DateTime startDate,
    required DateTime endDate,
  });

  /// Retrieves attendance records for a class within a date range.
  /// 
  /// Returns a list of [AttendanceRecordEntity] for the specified class.
  Future<Result<List<AttendanceRecordEntity>>> getAttendanceByClassAndDateRange({
    required int classId,
    required DateTime startDate,
    required DateTime endDate,
  });

  /// Updates an existing attendance record.
  /// 
  /// Returns updated [AttendanceRecordEntity] on success.
  /// Throws [ValidationException] on invalid data or [DataException] on update failure.
  Future<Result<AttendanceRecordEntity>> updateAttendanceRecord(
    AttendanceRecordEntity record,
  );

  /// Deletes an attendance record.
  /// 
  /// Returns success result on deletion,
  /// throws [DataException] if record not found or deletion fails.
  Future<Result<void>> deleteAttendanceRecord(int id);

  /// Retrieves attendance statistics for a class on a specific date.
  /// 
  /// Returns [AttendanceStatsEntity] for the specified class and date.
  Future<Result<AttendanceStatsEntity>> getAttendanceStatsForClass({
    required int classId,
    required DateTime date,
  });

  /// Retrieves attendance statistics for a student within a date range.
  /// 
  /// Returns [AttendanceStatsEntity] for the specified student.
  Future<Result<AttendanceStatsEntity>> getAttendanceStatsForStudent({
    required int studentId,
    required DateTime startDate,
    required DateTime endDate,
  });

  /// Retrieves attendance statistics for all classes on a specific date.
  /// 
  /// Returns a map of class IDs to their [AttendanceStatsEntity].
  Future<Result<Map<int, AttendanceStatsEntity>>> getAttendanceStatsForAllClasses({
    required DateTime date,
  });

  /// Exports attendance records for a class to Excel format.
  /// 
  /// [startDate] and [endDate] define the date range.
  /// Returns Excel data as a list of maps.
  Future<Result<List<Map<String, dynamic>>>> exportAttendanceToExcel({
    required int classId,
    required DateTime startDate,
    required DateTime endDate,
  });
}